### PR TITLE
WP 1138 Docker-based visual diff rendering

### PR DESCRIPTION
**Pull request by mmcgahan**
This PR sets up a Docker container that will run the visual diff tests on swarm-components and copy the `__diff_output__` results, if any, back to the local file system. From the developer's perspective, the behavior aligns with running the visual diffs locally, but since they are generated from a statically-defined environment (the Docker node base image - Ubuntu), they will be consistent across dev and CI environments.

Current downside:

- Running locally can take a long time because the node_modules have to be re-downloaded and the packages have to be-rebuilt every time there is a code change.

Optimizations:

- try to load the yarn cache into the Docker container on startup so that it doesn't have to download the packages each time it runs
- determine whether the `yarn install` layer of the container can be cached only for dependency changes, not source code changes. Currently there are some [challenges doing that](https://github.com/yarnpkg/yarn/issues/5703) with Yarn workspaces, but there might be some small wins possible.
- Better log output - it gets a little verbose

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,6 +6,10 @@
   ],
   "scripts": {
     "build": "lerna run build",
+    "build:visual": "docker build -t screenshot:latest ./;",
+    "test:visual": "yarn build:visual && docker run -t screenshot:latest yarn test:integration",
+    "copy:test:visual": "echo 'do something to get __diff_output__ out of the container'",
```
Comment (sergeychernyshev) [inline package.json]: Not sure if you pushed latest changes, but I don't see `__diff_output__` folder locally after diffs were written in Docker.

File: package.json [hunk 2]
```diff
@@ -6,6 +6,10 @@
...
+    "test:visual:execute": "cd packages/swarm-components/ && yarn test || true && ls src/__image_snapshots__",
```
Comment (sergeychernyshev) [inline package.json]: Not sure what this does - seems like it runs it locally do we need this?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+
+# Run everything after as non-privileged user.
+# USER pptruser
+
+RUN yarn build --scope=@meetup/swarm-styles --scope=@meetup/swarm-constants
+# RUN screenshot tests in /packages/swarm-components
+
+# RUN cd packages/swarm-components/ && yarn test || true && ls src/__image_snapshots__
+# copy diffs back to local, if any
```
Comment (sergeychernyshev) [inline Dockerfile]: Missing code?

File: package.json [hunk 4]
```diff
@@ -6,6 +6,10 @@
   ],
   "scripts": {
     "build": "lerna run build",
+    "build:visual": "docker build -t screenshot:latest ./;",
+    "test:visual": "yarn build:visual && docker run -t screenshot:latest yarn test:integration",
```
Comment (sergeychernyshev) [inline package.json]: We do need a way to generate initial screenshots, locally just running `yarn test` with empty `__image_snapshots__` generates the files.

Also need to be able to update screenshots with `yarn test:updateScreenshots` or something - locally, `yarn test -u` accomplishes that.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+# copy repo root (without node_modules) to working dir
+# node_modules are excluded by the .dockerignore rules
+# install dependencies
+
+COPY package.json ./
+COPY ./ ./
+RUN yarn install
+
+# # Add user so we don't need --no-sandbox.
+# # same layer as npm install to keep re-chowned files from using up several hundred MBs more space
```
Comment (sadafie) [inline Dockerfile]: There's a couple of commented out sections in this Dockerfile. Are these things that we will need? If not, can we remove?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+#   && mkdir -p /home/pptruser/Downloads \
+#   && chown -R pptruser:pptruser /home/pptruser \
+#   && chown -R pptruser:pptruser /node_modules \
+#   && chown -R pptruser:pptruser /packages
+
+
+# Run everything after as non-privileged user.
+# USER pptruser
+
+RUN yarn build --scope=@meetup/swarm-styles --scope=@meetup/swarm-constants
```
Comment (sadafie) [inline Dockerfile]: Were you able to figure out how to run the tests within the Dockerfile, or do we need to use `docker run`?

File: package.json (hunk 4 above)
Comment (sadafie) [reply package.json]: >We do need a way to generate initial screenshots, locally just running yarn test with empty __image_snapshots__ generates the files.

Not sure what you mean here. `__image_snapshots__` is tracked in git, so in any environment (local or Travis) will have an __image_snapshots__ directory.

>Also need to be able to update screenshots with yarn test:updateScreenshots or something - locally, yarn test -u accomplishes that.

The last two commands below this are `yarn test` and `yarn test:integration`:

https://github.com/meetup/swarm-ui/pull/72/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R13

To update you don't need a separate command, just append `-u`, like `yarn test:integration -u` at the root level or within `packages/swarm-components`. This is true in the current master.

File: package.json (hunk 2 above)
Comment (sadafie) [reply package.json]: I believe certain steps like copying and executing are now `docker run` commands instead of within the docker file, in which case, some of these scripts might not be needed anymore. However, we'll need the instructions for `docker run`.

**Comment by sadafie** [comment]
@mmcgahan Could you add instructions on how to run locally in the root level README? I believe part of the confusion with this PR is that there are some `docker run` commands that aren't documented.

**Comment by sadafie** [comment]
@mmcgahan I'm guessing you just haven't gotten to it yet, but we'll also need to update `.travis.yml` with the docker commands.

File: package.json (hunk 4 above)
Comment (sergeychernyshev) [reply package.json]: These commands run locally, not in Docker so they generate screenshots using developer computer's Chrome and not the same rendering engine that Docker version will use which will cause tests to fail.

We need Docker version to always be ran, for initial generation, for screenshot updates if developer made a change and needs a new screenshot generated and for actual testing.

Comment (sadafie) [reply package.json]: Sorry, I should have clarified. I meant you can run those yarn commands inside the docker container. I forget the exact command - but you can do the equivalent of ssh-ing into the container and run the yarn commands there and then use the copy command to move it to the local filesystem.

Or perhaps there is a `docker run` equivalent with something like `docker run yarn test:integration -u` with the correct container options passed in, but I'm not 100% sure that's possible. 

tl;dr those yarn commands already exist, we just need to document the correct docker steps.

**Comment by mmcgahan** [comment]
yes, sorry I should have left more notes, but wanted to get something into PR before Monday. I need to update the screenshots and do a little more cleanup, but will also add documentation so the process is super clear

File: package.json (hunk 1 above)
Comment (mmcgahan) [reply package.json]: Fixed by changing the `docker run` calls to mount the local directory directly in the container to keep them in sync

File: package.json (hunk 2 above)
Comment (mmcgahan) [reply package.json]: no longer needed/used

File: package.json (hunk 4 above)
Comment (mmcgahan) [reply package.json]: we now have 2 `make` commands to

1. run tests: `make test-visual`
2. update the local screenshots: `make test-visual-update`

Both commands update the local filesystem with screenshot changes generated inside the container

File: Dockerfile (hunk 5 above)
Comment (mmcgahan) [reply Dockerfile]: removed

File: Dockerfile (hunk 6 above)
Comment (mmcgahan) [reply Dockerfile]: I've set it up now so that the container just sets up the filesystem with the built assets, and then you use `docker run` to run tests and/or update snapshots.

**Comment by mmcgahan** [comment]
Travis renders some things with slightly-darker backgrounds:

toggle-test-jsx-toggle-should-match-screenshot-1-diff
